charmhelpers/contrib/hahelpers/apache.py

```diff
@@ -64,6 +70,36 @@ def get_cert(cn=None):
     return (cert, key)
 
 
+def validate_cert(cert, key, ca):
+    """
+    cert (bytes): cert.crt file contents (PEM armored)
+    key (bytes): cert.key file contents
+    ca (Optional[bytes]): None or private ca contents
+
+    return True if ok
+    otherwise return False, set blocked status, and set useful error message
+    """
+
+    context = None
+    if ca:
+        context = ValidationContext(trust_roots=[ca])
+
+    # split certs so we can pass leaf and intermediate certs separately
+    # The leaf/server certificate _must_ be the first cert https://www.rfc-editor.org/rfc/rfc5246#page-48
+    certs = re.findall(r'-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----', cert.decode(), re.DOTALL)
+    certs = list(map(lambda x: x.encode(), certs))
+
+    validator = CertificateValidator(certs[0], certs[1:], validation_context=context)
+
+    try:
+        validator.validate_usage({"digital_signature"}, {"server_auth"})
+    except Exception as e:
+        status_set(WORKLOAD_STATES.BLOCKED, str(e))
```

Comment:
Probably this status_set has only a temporary effect. Most of the OpenStack charms call assess_status() or equivalent at the end of the hook execution and that will override any status set during the hook execution.

How to set the status appropriately might be a next step, but logging an error or warning is definitely one step forward than the current status.

Reply:
For example, charm-keystone has `assess_status()` here at the end of a hook execution:
https://github.com/openstack/charm-keystone/blob/9b2f7e545dd682e90d5403a8cb8145309c42d5cf/hooks/keystone_hooks.py#L933
and eventually calls:
https://github.com/juju/charm-helpers/blob/8b22c53d86cb1ba745106e12ac713cff6810dd1e/charmhelpers/contrib/openstack/utils.py#L1003-L1010